extmod/modussl_mbedtls.c

```diff
@@ -121,12 +123,29 @@ STATIC mp_obj_ssl_socket_t *socket_new(mp_obj_t sock, struct ssl_args *args) {
     o->base.type = &ussl_socket_type;
     o->sock = sock;
 
+    if ((args->psk_ident.u_obj != mp_const_none || args->psk_key.u_obj != mp_const_none) &&
+            (args->cert.u_obj != mp_const_none || args->key.u_obj != mp_const_none)) {
+        mp_raise_ValueError("cannot have cert&key and psk_ident&psk_key");
+    }
+    uint8_t psk_mode = args->psk_ident.u_obj != mp_const_none;
+    if (psk_mode) {
+        if (args->server_side.u_bool) {
+            mp_raise_ValueError("PSK ciphers only supported for client connections");
+        }
+    }
+
     int ret;
     mbedtls_ssl_init(&o->ssl);
     mbedtls_ssl_config_init(&o->conf);
-    mbedtls_x509_crt_init(&o->cacert);
-    mbedtls_x509_crt_init(&o->cert);
-    mbedtls_pk_init(&o->pkey);
+    if (psk_mode) {
+        memset(&o->cacert, 0, sizeof(o->cacert));
```

Comment:
Are these `memset`s needed? I believe mbedtls is smart enough to keep away from these objects when in PSK mode.

Reply:
I ended up doing this to avoid having to add a "mode" field so the `free` function knows which to call mbedtls free on and which not. I briefly looked at the MP heap allocator and I do not believe that is zeroes allocated memory, or at least there's a #define there, which means it's not guaranteed. And it wouldn't be good for the mbedtls free functions to be called on some random struct. I'm happy for suggestions if this too weird.

Reply:
It's a bad practice to call compound `_free` functions without initializing those first, anyway (whether it's zeros or not).

I think it's too weird and a mode variable should be added. (Or instead, use some inner variable of mbedtls, which probably keeps this state itself)